File: SYSNETPROJECTS/sysnet2/prj1/23 Thursday/TCPclient.java

/*
 * TCPclient.java
 * Systems and Networks II
 * Project 1
 *
 * This file describes the functions to be implemented by the TCPclient class
 * You may also implement any auxillary functions you deem necessary.
 */
import java.net.*;
import java.io.*;
import java.util.Scanner;

public class TCPclient
{
	private Socket _socket; // the socket for communication with a server
    private OutputStream sendStream;
    private InputStream recvStream;
    
   
       
	/**
	 * Constructs a TCPclient object.
	 */
	public TCPclient()
	{
      
	}

	/**
	 * Creates a streaming socket and connects to a server.
	 *
	 * @param host - the ip or hostname of the server
	 * @param port - the port number of the server
	 *
	 * @return - 0 or a negative number describing an error code if the connection could not be established
	 */
	public int createSocket(String host, int port)
	{
        try{
        _socket = new Socket(host, port);
        sendStream = _socket.getOutputStream();
        recvStream = _socket.getInputStream();
        }catch(Exception ex){
        	System.out.println("Socket stream Fail:");
        	return -1;
        }
        return 0;
	}
    
	/**
	 * Sends a request for service to the server. Do not wait for a reply in this function. This will be
	 * an asynchronous call to the server.
	 *
	 * @request - the request to be sent
	 *
	 * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int sendRequest(String request)
	{
	if(request.length() == 0)return -1;
        try{
            byte[] sendBuff = new byte[request.length()];
            sendBuff = request.getBytes();
            sendStream.write(sendBuff, 0, sendBuff.length);
        }catch(IOException ex){
            System.err.println("IOException in sendRequst");
        	return -1;
	}
		return 0;
	}

	/**
	 * Receives the server's response. Also displays the response in a
	 * clear and concise format.
	 *
	 * @return - the server's response or NULL if an error occured
	 */
	public String receiveResponse()
	{
		String response = "";		
		try{
            int dataSize;
            while((dataSize = recvStream.available()) == 0);
            byte[] recvBuff = new byte[dataSize];
            recvStream.read(recvBuff, 0,dataSize);
            response = new String(recvBuff,0,dataSize);
        }catch (Exception ex){
            System.err.println("IOException in sendResponse");
        	return null;
	}
		return response;
	}
	
	/*
     * Prints the response to the screen in a formatted way.
     *
     * response - the server's response as an XML formatted string
     *
     */
	public static void printResponse(String response)
	{
		System.out.println("Response from server: " + response);  
	}
    
    
	/*
	 * Closes an open socket.
	 *
     * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int closeSocket()
	{
		try{    
		    sendStream.close();  
		    recvStream.close();    	
		    _socket.close();    
		}catch(Exception ex){  
		    System.err.println("IOException in close"); 
			return -1;
		}   
		return 0;
	}
    
	
	/**
	 * The main function. Use this function for
	 * testing your code. We will use our own main function for testing.
	 */
	public static void main(String[] args)
	{
		String response ="";
		if (args.length > 0) {	 
			final String servName = args[0];
		    final int servPort = Integer.parseInt(args[1]);	
		    TCPclient client = new TCPclient();  
		    client.createSocket(servName, servPort); 
		    
		   if(client.sendRequest("<echo>sfglk</echo>") >= 0 ){    
		    response = client.receiveResponse();    
		    client.printResponse(response);    
		}	  
		   if(client.sendRequest("") >= 0 ){  
		    response = client.receiveResponse();    
		    client.printResponse(response);    
		}
		    
		   if(client.sendRequest("<echo>sf\nglk</echo>") >= 0 ){    
		    response = client.receiveResponse();    
		    client.printResponse(response);
			}    
		
	  	    client.closeSocket();    
		}
	}
    
}

File: SYSNETPROJECTS/sysnet2/prj1/16 Thursday/TCPclient.c
/*
 * TCPclient.c
 * Systems and Networks II
 * Project 1
 *
 * This file implements the functions for the TCPclient.
 * and implements many auxillary functions we deem necessary.
 */

 //includes
 #include "TCPclient.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netdb.h> 


//???????????includes
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#include <unistd.h>


/*
 * Creates a streaming socket and connects to a server.
 *
 * serverName - the ip address or hostname of the server given as a string
 * port       - the port number of the server
 * dest       - the server's address information; the structure should be created with information
 *              on the server (like port, address, and family) in this function call
 *
 * return value - the socket identifier or a negative number indicating the error if a connection could not be established
 */
int createSocket(char * serverName, int port, struct sockaddr_in ** dest){
  int sockfd;
  struct hostent *hostptr;
  sockfd = socket(AF_INET, SOCK_STREAM, 0); //creates tcp socket
  hostptr = gethostbyname(serverName); // getting information from host 
  return sockfd;
}

/*
 * Sends a request for service to the server. This is an asynchronous call to the server, 
 * so do not wait for a reply in this function.
 * 
 * sock    - the socket identifier
 * request - the request to be sent encoded as a string
 * dest    - the server's address information
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int sendRequest(int sock, char * request, struct sockaddr_in * dest){

}

/*
 * Receives the server's response formatted as an XML text string.
 *
 * sock     - the socket identifier
 * response - the server's response as an XML formatted string to be filled in by this function
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int receiveResponse(int sock, char ** response){
}

/*
 * Prints the response to the screen in a formatted way.
 *
 * response - the server's response as an XML formatted string
 *
 */
void printResponse(char* response){

}

/*
 * Closes the specified socket
 *
 * sock - the ID of the socket to be closed
 * 
 * return - 0, if no error; otherwise, a negative number indicating the error
 */
int closeSocket(int sock){


]
